Source/WebCore/platform/graphics/cocoa/ImageUtilitiesCocoa.mm

```diff
@@ -0,0 +1,95 @@
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#import "config.h"
+#import "ImageUtilities.h"
+
+#if PLATFORM(MAC)
+#import <pal/spi/mac/NSImageSPI.h>
+#else
+#import <pal/spi/cocoa/CoreSVGSPI.h>
+#endif
+
+#if !PLATFORM(MAC)
+SOFT_LINK_PRIVATE_FRAMEWORK(CoreSVG)
+SOFT_LINK(CoreSVG, CGSVGDocumentCreateFromData, CGSVGDocumentRef, (CFDataRef data, CFDictionaryRef options), (data, options))
+SOFT_LINK(CoreSVG, CGSVGDocumentRelease, void, (CGSVGDocumentRef document), (document))
+#endif
+
+namespace WebCore {
+
+bool canDataBeDecodedAsSVG(std::span<const uint8_t> data)
+{
+    RetainPtr nsData = toNSData(data);
+#if PLATFORM(MAC)
+    static Class NSSVGImageRep = NSClassFromString(@"_NSSVGImageRep");
+    if (!NSSVGImageRep)
+        return false;
+
+    return [NSSVGImageRep canInitWithData:nsData.get()];
+#else
+    RetainPtr parser = adoptNS([[NSXMLParser alloc] initWithData:nsData.get()]);
+    if (!parser || ![parser parse])
+        return false;
+
+    CGSVGDocumentRef document = CGSVGDocumentCreateFromData((__bridge CFDataRef)nsData.get(), nullptr);
+    bool result = !!document;
+    CGSVGDocumentRelease(document);
+    return result;
+#endif
+}
+
+std::optional<Vector<FloatSize>> availableSizesFromData(std::span<const uint8_t> data, bool isSVGData)
+{
+    if (isSVGData)
+        return Vector<FloatSize> { };
+
+    RetainPtr nsData = toNSData(data);
+    auto source = adoptCF(CGImageSourceCreateWithData((__bridge CFDataRef)nsData.get(), nullptr));
+    if (!source)
+        return std::nullopt;
+
+    size_t count = CGImageSourceGetCount(source.get());
+    Vector<FloatSize> sizes;
+    sizes.reserveInitialCapacity(count);
+    for (size_t index = 0; index < count; ++index) {
+        auto properties = adoptCF(CGImageSourceCopyPropertiesAtIndex(source.get(), index, nullptr));
+        if (!properties) {
+            sizes.append(FloatSize { });
+            continue;
+        }
+
+        float width = 0, height = 0;
+        CFNumberRef cfWidth = (CFNumberRef)CFDictionaryGetValue(properties.get(), kCGImagePropertyPixelWidth);
```

Comment:
We have `checked_cf_cast`, which should be used in cases like this.

Reply:
This part is removed.